File: Java_MVVM/app/src/main/java/eu/gotoinc/requesinjava_mvvm/util/architecture/Factory.java
package eu.gotoinc.requesinjava_mvvm.util.architecture;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import eu.gotoinc.requesinjava_mvvm.repository.ApiContract;
import eu.gotoinc.requesinjava_mvvm.viewmodel.HomeScreenViewModel;

public class Factory extends ViewModelProvider.NewInstanceFactory {

	private ApiContract api;

	public Factory(ApiContract api) {
		super();
		this.api = api;
	}



	@NonNull
	@Override
	public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
		if (modelClass == HomeScreenViewModel.class) {
			return (T) new HomeScreenViewModel(api);
		}
		return null;
	}
}

File: Kotlin_MVVM/app/src/main/java/eu/gotoinc/kotlinmvvm/util/architecture/Factory.kt
@file:Suppress("UNCHECKED_CAST")

package eu.gotoinc.kotlinmvvm.util.architecture

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import eu.gotoinc.kotlinmvvm.repository.ApiContract
import eu.gotoinc.kotlinmvvm.viewmodel.HomeScreenViewModel

class Factory(private val api: ApiContract) : ViewModelProvider.Factory {

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        return when(modelClass){
            HomeScreenViewModel::class.java -> HomeScreenViewModel(api) as T
            else -> throw IllegalArgumentException("ViewModel Not Found")
        }
    }

}